File: python/wallet/flash_provider.py
from wallet.tron_provider import TronProvider
from tronpy import AsyncTron
from tronpy.providers import AsyncHTTPProvider
from typing import Optional, Any
import httpx
import os
import json


class FlashProvider(TronProvider):
    def __init__(
        self,
        rpc_url: Optional[str] = None,
        api_key: Optional[str] = None,
        privy_app_id: Optional[str] = None,
        privy_app_secret: Optional[str] = None,
        wallet_id: Optional[str] = None,
        keystore_path: Optional[str] = None,
        keystore_password: Optional[str] = None,
        keystore=None,
    ):
        """
        Initialize FlashProvider with Privy integration.
        After construction, call `await init()` to load credentials from keystore.
        Or use `await FlashProvider.create(...)` for one-step setup.
        """
        super().__init__(
            rpc_url=rpc_url,
            api_key=api_key,
            private_key=None,
            keystore_path=keystore_path,
            keystore_password=keystore_password,
            keystore=keystore,
        )

        self.privy_app_id = privy_app_id or os.getenv("PRIVY_APP_ID")
        self.privy_app_secret = privy_app_secret or os.getenv("PRIVY_APP_SECRET")
        self.wallet_id = wallet_id or os.getenv("PRIVY_WALLET_ID")

        if self.wallet_id:
            self.address = self.wallet_id

        if not self.privy_app_id or not self.privy_app_secret or not self.wallet_id:
            print("Warning: Privy credentials (APP_ID, APP_SECRET, WALLET_ID) not fully provided.")

    async def init(self) -> "FlashProvider":
        """Load additional Privy credentials from keystore.
        Keystore keys: privyAppId, privyAppSecret, walletId
        """
        await super().init()

        ks_privy_app_id = self.keystore.get("privyAppId")
        ks_privy_app_secret = self.keystore.get("privyAppSecret")
        ks_wallet_id = self.keystore.get("walletId")

        if not self.privy_app_id and ks_privy_app_id:
            self.privy_app_id = ks_privy_app_id
        if not self.privy_app_secret and ks_privy_app_secret:
            self.privy_app_secret = ks_privy_app_secret
        if not self.wallet_id and ks_wallet_id:
            self.wallet_id = ks_wallet_id
            self.address = self.wallet_id

        return self

    @classmethod
    async def create(
        cls,
        rpc_url: Optional[str] = None,
        api_key: Optional[str] = None,
        privy_app_id: Optional[str] = None,
        privy_app_secret: Optional[str] = None,
        wallet_id: Optional[str] = None,
        keystore_path: Optional[str] = None,
        keystore_password: Optional[str] = None,
        keystore=None,
    ) -> "FlashProvider":
        """Factory: create and init a FlashProvider in one step."""
        provider = cls(
            rpc_url=rpc_url,
            api_key=api_key,
            privy_app_id=privy_app_id,
            privy_app_secret=privy_app_secret,
            wallet_id=wallet_id,
            keystore_path=keystore_path,
            keystore_password=keystore_password,
            keystore=keystore,
        )
        await provider.init()
        return provider

    async def sign_transaction(self, transaction: Any) -> Any:
        """Sign transaction using Privy API."""
        tx_id = transaction.txid

        sign_url = f"https://auth.privy.io/api/v1/wallets/{self.wallet_id}/sign"

        headers = {
            "Authorization": f"Basic {self._get_basic_auth()}",
            "Content-Type": "application/json",
            "privy-app-id": self.privy_app_id,
        }

        payload = {
            "method": "raw_sign",
            "params": {"message": tx_id, "encoding": "hex"},
        }

        async with httpx.AsyncClient() as client:
            resp = await client.post(sign_url, json=payload, headers=headers)
            resp.raise_for_status()
            data = resp.json()

            if "signature" not in data:
                raise ValueError("Privy signing response did not contain a signature")
            transaction._signature = [data["signature"]]
            return transaction

    async def sign_message(self, message: bytes) -> str:
        """Sign an arbitrary message and return a raw signature string.

        If Privy credentials are not configured, falls back to local signing
        (same as TronProvider).
        """
        if not self.privy_app_id or not self.privy_app_secret or not self.wallet_id:
            return await super().sign_message(message)

        msg_hex = message.hex()

        sign_url = f"https://auth.privy.io/api/v1/wallets/{self.wallet_id}/sign"
        headers = {
            "Authorization": f"Basic {self._get_basic_auth()}",
            "Content-Type": "application/json",
            "privy-app-id": self.privy_app_id,
        }
        payload = {"method": "raw_sign", "params": {"message": msg_hex, "encoding": "hex"}}

        async with httpx.AsyncClient() as client:
            resp = await client.post(sign_url, json=payload, headers=headers)
            resp.raise_for_status()
            data = resp.json()
            if "signature" not in data:
                raise ValueError("Privy signing response did not contain a signature")
            return data["signature"]

    
    def _get_basic_auth(self) -> str:
        import base64

        creds = f"{self.privy_app_id}:{self.privy_app_secret}"
        return base64.b64encode(creds.encode()).decode()

    async def send_transaction(self, to_address: str, amount: float) -> dict:
        """Send a flash transaction using Privy signing."""
        if not self.address:
            raise ValueError("Address not available for signing")
        txn = (
            self.client.trx.transfer(self.address, to_address, amount)
            .memo("Privy Flash Transaction")
            .fee_limit(100_000_000)
        )
        txn = await txn.build()
        signed_txn = await self.sign_transaction(txn)
        result = await signed_txn.broadcast()
        return result

File: typescript/src/wallet/flash_provider.ts
import { TronProvider, type TronProviderOptions } from './tron_provider';
import { TronWeb } from 'tronweb';

export interface FlashProviderOptions extends TronProviderOptions {
  flashNode?: string;
  privyAppId?: string;
  privyAppSecret?: string;
  walletId?: string;
}

export class FlashProvider extends TronProvider {
  protected flashTronWeb: any;
  protected privyAppId: string;
  protected privyAppSecret: string;
  protected walletId: string;

  constructor(opts: FlashProviderOptions = {}) {
    super(opts);

    // privyAppId/privyAppSecret/walletId come from keystore (or explicit opts), not from env.
    this.privyAppId = opts.privyAppId || '';
    this.privyAppSecret = opts.privyAppSecret || '';
    this.walletId = opts.walletId || '';

    // Load missing Privy credentials from keystore in constructor (best-effort).
    if (typeof (this.keystore as any).getSync === 'function') {
      const ksPrivyAppId = (this.keystore as any).getSync('privyAppId') as string | undefined;
      const ksPrivyAppSecret = (this.keystore as any).getSync('privyAppSecret') as string | undefined;
      const ksWalletId = (this.keystore as any).getSync('walletId') as string | undefined;
      if (!this.privyAppId && ksPrivyAppId) this.privyAppId = ksPrivyAppId;
      if (!this.privyAppSecret && ksPrivyAppSecret) this.privyAppSecret = ksPrivyAppSecret;
      if (!this.walletId && ksWalletId) this.walletId = ksWalletId;
    }

    if (this.walletId) {
      this.address = this.walletId;
    }

    const flashNode = opts.flashNode || process.env.TRON_FLASH_RPC_URL || opts.fullNode || process.env.TRON_RPC_URL || 'https://api.trongrid.io';
    const fullNode = opts.fullNode || process.env.TRON_RPC_URL || 'https://api.trongrid.io';

    if (flashNode !== fullNode) {
      const ksPrivateKey =
        typeof (this.keystore as any).getSync === 'function'
          ? ((this.keystore as any).getSync('privateKey') as string | undefined)
          : undefined;
      const tronOpts: any = {
        fullHost: flashNode,
        privateKey: opts.privateKey || ksPrivateKey || undefined,
      };
      this.flashTronWeb = new TronWeb(tronOpts);
    } else {
      this.flashTronWeb = this.tronWeb;
    }
  }

  /**
   * Load additional Privy credentials from keystore.
   * Keystore keys: privyAppId, privyAppSecret, walletId
   */
  async init(): Promise<this> {
    // Compatibility: constructor already loads credentials from keystore when possible.
    // Still read again to support custom keystore implementations.
    await super.init();
    return this;
  }

  /**
   * Factory: create and init a FlashProvider in one step.
   */
  static async create(opts: FlashProviderOptions = {}): Promise<FlashProvider> {
    const provider = new FlashProvider(opts);
    await provider.init();
    return provider;
  }

  override async signMessage(message: Uint8Array): Promise<string> {
    // If Privy not configured, fall back to local signing (TronProvider)
    if (!this.privyAppId || !this.privyAppSecret || !this.walletId) {
      return super.signMessage(message);
    }

    const msgHex = Buffer.from(message).toString("hex");
    const url = `https://auth.privy.io/api/v1/wallets/${encodeURIComponent(this.walletId)}/sign`;
    const auth = btoa(`${this.privyAppId}:${this.privyAppSecret}`);

    const response = await fetch(url, {
      method: "POST",
      headers: {
        Authorization: `Basic ${auth}`,
        "Content-Type": "application/json",
        "privy-app-id": this.privyAppId,
      },
      body: JSON.stringify({
        method: "raw_sign",
        params: {
          message: msgHex,
          encoding: "hex",
        },
      }),
    });

    if (!response.ok) {
      const errBody = await response.text().catch(() => "");
      throw new Error(
        `Privy signing failed (${response.status}): ${errBody || response.statusText}`,
      );
    }

    const data = await response.json();
    const signature = data.signature;
    if (!signature) {
      throw new Error("Privy signing response did not contain a signature");
    }
    return signature;
  }

  async sign(transaction: any): Promise<any> {
    if (!this.privyAppId || !this.privyAppSecret || !this.walletId) {
      return super.sign(transaction);
    }

    const txID = transaction.txID;

    const url = `https://auth.privy.io/api/v1/wallets/${encodeURIComponent(this.walletId)}/sign`;
    const auth = btoa(`${this.privyAppId}:${this.privyAppSecret}`);

    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Authorization': `Basic ${auth}`,
        'Content-Type': 'application/json',
        'privy-app-id': this.privyAppId
      },
      body: JSON.stringify({
        method: 'raw_sign',
        params: {
          message: txID,
          encoding: 'hex'
        }
      })
    });

    if (!response.ok) {
      const errBody = await response.text().catch(() => '');
      throw new Error(`Privy signing failed (${response.status}): ${errBody || response.statusText}`);
    }

    const data = await response.json();
    const signature = data.signature;

    if (!signature) {
      throw new Error('Privy signing response did not contain a signature');
    }

    const signedTx = JSON.parse(JSON.stringify(transaction));
    if (!signedTx.signature) {
      signedTx.signature = [];
    }
    signedTx.signature.push(signature);

    return signedTx;
  }

  async sendTransaction(toAddress: string, amount: number): Promise<any> {
    if (!this.address) throw new Error("Address not available for signing");
    try {
      const tradeobj = await this.flashTronWeb.transactionBuilder.sendTrx(toAddress, amount, this.address);
      const signedtxn = await this.sign(tradeobj);
      const receipt = await this.flashTronWeb.trx.sendRawTransaction(signedtxn);
      return receipt;
    } catch (error) {
      console.warn("Flash transaction failed", error);
      throw error;
    }
  }
}
